### PR TITLE
Web application responsive & color selector | Code clean

### PR DESCRIPTION
**Pull request by gottburgm**
### **General** 
  - Cleaned as much as i could the html/css and JavaScript hardcoded stuff. I moved all the JavaScript code in **wardemo/src/main/webapp/js/script.js** which is included by the html.
  - Added responsive HTML/CSS support to fit on mobile devices .
  - Added a color selector which allow the user(s) to choose a color and a led to fill with this color from the web application .
 - **WANRNING:** The HTML/CSS or JavaScript may contain small errors as i couldn't test the last changes that i made on the raspberry. It is working on my browser but need to be compiled and tested on a raspberry.

### **Directories** 
  - **wardemo/src/main/webapp/css** for the CSS files .
  - **wardemo/src/main/webapp/js** for the JavaScript files .

### **Code**
  - Added HTML headers to follow W3 rules
  - Added **style.css** for custom CSS rules
  - Added **script.js** which contains all the JavaScript functions
  - Removed hardcoded IP addresses / references
  
  ```java
    // Removed from FrameBuffer.java
   out.println("var request = new XMLHttpRequest();");
   out.println("var theUrl = \"http://10.0.0.203:8080/demo-1.0-SNAPSHOT/FrameBuffer?i=\" + " + i + " + \"&j=\" + " + j + ";");
   out.println("request.open( \"GET\", theUrl, false );");
  ```

  ```javascript
   var client = new XMLHttpRequest();
   var requestUrl = window.location.protocol + "//" + window.location.host + "/" + window.location.pathname + "?i=" + event.target.i + "&j=" + event.target.j + "&red=" + getRed(userColor.value) + "&green=" + getGreen(userColor.value) + "&blue=" + getBlue(userColor.value);

   client.open("GET", requestUrl);
   client.send();
  ```
  
  ```
   The request URL looks like:
   http://<RASPBERRY_IP>/<FILE>?i=<CORD_X>&j=<CORD_Y>&red=<RED_HEX_CODE>&green=<GREEN_HEX_CODE>&blue=<BLUE_HEX_CODE>)

   => The RGB value of the elected color is splitted (R, G and B) and sent with the coords of the pixel clicked (so the pixel on which the color should be applied). The page will return the new "pixel mapping"
  ```

  - Removed all debugging stuff
  - Added the correction of the RGB565 function:
  
  ```java
    // return (short) (blue*256+red*16+green/8+(green%8)*512);
    return (short) ((blue<<8)+red*8+green/8+((green%8)<<13));
  ```

  - Added predefined colors because some colors are impossible to make with the leds of the Raspberry PI. Those colors are defined in **script.js** : 

  ```javascript
   colors: {
     'tetrad1': 'rgb(255, 0, 6)',
     'tetrad2': 'rgb(27, 255, 0)',
     'tetrad3': 'rgb(225, 255, 0)',
     'tetrad4': 'rgb(255, 116, 0)',
     'tetrad5': 'rgb(0, 1, 255)',
     'tetrad6': 'rgb(228, 0, 255)',
     'tetrad7': 'rgb(0, 255, 19)',
     'tetrad8': 'rgb(255, 255, 255)',
     'tetrad9': 'rgb(0, 0, 0)'
   },
  ```

### TODO
  - [ ] Test this commit on the Raspberry PI
  - [ ] Test the "responsivility" from different mobile devices 
  - [ ] Read all the HTML which is hardcoded in **FrameBuffer.java** from a template file
  - [ ] Remove useless things from the JavaScript code
  - [ ] Clean the JavaScript code
  - [ ] Use POST requests instead of GET requests

### **Librairies & References** 
  -  Responsive Design & CSS: [Bootstrap Website](https://getbootstrap.com/)
  - Color selection: 
    - [Bootstrap Colorpicker](https://github.com/farbelous/bootstrap-colorpicker.git)
    - [Bootstrap Colorpicker Examples](https://farbelous.io/bootstrap-colorpicker/tutorial-Basics.html)

### **Contact**
  - Github: [gottburgm](https://github.com/gottburgm/)
  - Email: <a href="mailto:mgottbur@redhat.com">mgottbur@redhat.com</a>

### CODE REVIEW

File: wardemo/src/main/java/org/jfclere/tomcatPI/FrameBuffer.java [hunk 1]
```diff
@@ -23,114 +23,101 @@
...
+        String param_i = request.getParameter("i");
+        String param_j = request.getParameter("j");
+        String param_red = request.getParameter("red");
+        String param_green = request.getParameter("green");
+        String param_blue = request.getParameter("blue");
+
+        if (param_j != null && param_i != null) {
+            int p_i = Integer.parseInt(param_i);
+            int p_j = Integer.parseInt(param_j);
+            int red = Integer.parseInt(param_red);
```
Comment (jfclere) [inline wardemo/src/main/java/org/jfclere/tomcatPI/FrameBuffer.java]: What happens is red, green or blue are null?

File: wardemo/src/main/java/org/jfclere/tomcatPI/PIFrameBuffer.java [hunk 2]
```diff
@@ -59,63 +58,11 @@ public int getGreen(short pixel) {
 		cache = 0x0007;
 		return (pixel & cache) * 8 + green;
 	}
-
+
```
Comment (jfclere) [inline wardemo/src/main/java/org/jfclere/tomcatPI/PIFrameBuffer.java]: If you remove the code for main() why you didn't remove it?
We don't need it correct?

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
...GTgGnsZQWMYrcOARoIReAALBeWhf+RUCAIEsECFQHLkwR5zj4JW3t5WOUU5djvgQIawD53EDsctmYz8xGaZGPBUR3qNkiGwqDICUYIFpqBgRaayCfFiAWR2wWvoobmzxdF8N5kyxXmvap/sgGcLF/aoBosbG+lE395R8zCA4BqUYgOgYq+HtvBrT0LK15X8lZwx5f9klCX0rdgXzIIGbdhXMqZtHzJhuptEjmsFc4KzmN5IFPtfM7gWw2kPczSIqQSPUDYKYBMamsBCpKphW0iA5H8AbMDPJOQYjLZg1Vk4G49GlCYNYAkdOd0kwRQ8FCyAHydgLZ6Z2AqrVtjDUQ7hCEmrkEooDAsB2YnBCvkBpZ6yBvJpCd7Mn5zJ6C4QF2BUQPgHEIGUrGnHzQ8rlMekBeTyAzwDJksxwM4+w3BY02B8mIl0CmFRm+ZscxAuSnvwqQsECTIGSV6FEoJFTygVuzB5xAsKqBvAQE3+nkVoJDI1BJIaPBWik7ZSu5NIp5A3mRQaTFvLgkO9fVgEgMqqeVfb+p55tijWH+Kea71ubq4v8Sl8089sZKbKEZNq+VUfISJJF7j79WrbYgS994ZEf+nIz0pNFRWqapSmK6P45i3OQuItIiPDyg6RnxZ4D0g+CFPxAzluoRsWsaA6I6JOqVWCisDvJ0BgHTzMSRgMi0vmi8R+sR6tg/XUh7kCc7kMRqSNkTBDx0OkAUegFcMazciBXNpm798R6klXap/WZz49TQwBHqEcj4oCToUPjUuP9lfxcbyKMAwT6bTf1qqIIQDl3i5oCERNmVm0wgW4A8BGRxMX3hWh8bEV5Rvfp4DS5F3djWH2ztDNWKW7OBjgjIwsDWaKRknJjqMsh9QCa1p608lLovFkBE969DYtYelSzwSRcg535vAsFeNU9SzRCYZb4LDmxmFQKkwYGM+5y/G7b1uxMIylLdyE5yxIyYsoXWhQIpzQhYPi3JkJoKkB9+BxD0OMuyOEBe36DgyPSrxscmATldgKj8PxrkA/kA5PYMgkrocwIQ6GSRGmF0VaNqBKQZ5FYDEZSDzFTzq9mBQjAayE1A+ryDTzcQZe0Ibbxj7EwpAmTrJwEimZR9CCPtODhzxuNtY19Zd2Lf/fjCTnEiDAOg62j1utb/dv9mZ/aHCj4AyOHbsW3/As0BTzIgeJU7AAAAAElFTkSuQmCC");
```
Comment (jfclere) [inline wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: No sure that is right we need the original png file in the repo.

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+  background: #000;
+  border-top: 1px solid #fff;
+  position: absolute;
+  top: 0;
+  left: 0;
+  width: 100%;
+  margin-top: -1px; }
+
+.colorpicker-hue {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAABkCAMAAABw8qpSAAABLFBMVEXqFBb/ABH/ACL/ADH/AEH/AFD/AGD/AG7/AH7/AI3/AJ3/AKz/ALz/AMr/ANv/AOr/APr2AP/mAP/XAP/HAP+4AP+oAP+aAP+JAP97AP9rAP9cAP9MAP8+AP8tAP8fAP8PAP8BAv8AEP8AH/8AL/8APv8ATv8AXP8Abf8Ae/8Ai/8Amv8Aqv8AuP8Ayf8A1/8A5/8A9/8A//gA/+kA/9kA/8oA/7oA/6wA/5sA/40A/30A/24A/14A/1AA/z8A/zEA/yEA/xEB/wMN/wAd/wAs/wA8/wBK/wBb/wBp/wB5/wCI/wCY/wCm/wC3/wDF/wDV/wDk/wD1/wD/+gD/7AD/3AD/zAD/vgD/rQD/nwD/jgD/gAD/cAD/YgD/UQD/QwD/MgD/JAD/FAD4Eg42qAedAAAAh0lEQVR4XgXAg3EDAAAAwI9to7Zt27a1/w49BASFhEVExcQlJKWkZWTl5BUUlZRVVNXUNTS1tHXo1KVbj159+g0YNGTYiFFjxk2YNGXajFlz5i1YtGTZilVr1m3YtGXbjl179h04dOTYiVNnzl24dOXajVt37j149OTZi1dv3n349OXbj19//wOxE1dQ8reGAAAAAElFTkSuQmCC"); }
```
Comment (jfclere) [inline wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: Same as before.

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+.colorpicker-alpha {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAABkCAQAAAAVxWkcAAABr0lEQVR4Xo2VwU0DQQxF7dmRuNIFlzlSAR3QAaXQQdIBJVABFXDcOVAAd67cjJLR07dkhcSrkZKfb/t7bG88rFo3B5gZPMNycItu2xloGV7MWHzM9zuzFWCkmA0nK6AszCUJDW6+mG6R03ncw5v8EMTEvZ2O3AliYjpslblc0RF9LmZYWxURU6aKytWZYsoWCAe+xwOZp1GsEukGiIkYxcQCHck99+gRgB7JncyIB5SGEhP3Yh5P6JwX+u6AnYot104d8DJT7uH7M9JH6OZbimj0vfMVaYnJIZFJDBW9kHlerL2C6JV4mSt7uuo2N57RxnZ+usQjn0R1jwBJBrNO3evJpVYUWsJ/E3UiXRlv24/7YZ04xmEdWlzcKS+B/eapeyMvFd2k0+hRk/T0AmTW8h69s2sjYMsdPntECiILhAeIMZAeH4QvUwfn6ijC0tTV+fT9ky8jM9nK2g7Ly1VjSpKYq6IvsAm7MtNu1orEqa/K3KNvgMFdhfquPfJmp2dbh0/8Gzb6Y22ViaNr6n5410zXdngVhbu6XqdOtWOuin5hjABGp4a2uotZ71MVCfwDBt2/v37yo6AAAAAASUVORK5CYII=");
```
Comment (jfclere) [inline wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: Idem

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+  background: #ffffff;
+  position: absolute;
+  top: 0;
+  left: 0;
+  width: 1px;
+  border: none;
+  margin-top: 0; }
+
+.colorpicker.colorpicker-horizontal .colorpicker-hue {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAABCAMAAAAfBfuPAAABLFBMVEXqFBb/ABH/ACL/ADH/AEH/AFD/AGD/AG7/AH7/AI3/AJ3/AKz/ALz/AMr/ANv/AOr/APr2AP/mAP/XAP/HAP+4AP+oAP+aAP+JAP97AP9rAP9cAP9MAP8+AP8tAP8fAP8PAP8BAv8AEP8AH/8AL/8APv8ATv8AXP8Abf8Ae/8Ai/8Amv8Aqv8AuP8Ayf8A1/8A5/8A9/8A//gA/+kA/9kA/8oA/7oA/6wA/5sA/40A/30A/24A/14A/1AA/z8A/zEA/yEA/xEB/wMN/wAd/wAs/wA8/wBK/wBb/wBp/wB5/wCI/wCY/wCm/wC3/wDF/wDV/wDk/wD1/wD/+gD/7AD/3AD/zAD/vgD/rQD/nwD/jgD/gAD/cAD/YgD/UQD/QwD/MgD/JAD/FAD4Eg42qAedAAAAbUlEQVR4XgXAghEDsbxtlrZt27ax/w49ACAYQTGcICmaYTleECVZUTXdMC1Wm93hdLk9Xp8/EAyFI9FYPJFMpTPZXL5QLJUr1Vq90Wy1O91efzAcjSfT2XyxXK03293+cDydL9fb/fF8vT/f3x+LfRNXARMbCAAAAABJRU5ErkJggg=="); }
```
Comment (jfclere) [inline wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: sure?

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+.colorpicker.colorpicker-horizontal .colorpicker-alpha {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAKCAQAAADoFTP1AAAB9ElEQVR4XoWTQW4VMRBEu9qWEimL7DhEMp8NF+ASnJJLcAQgE1bcgBUSkYKUuHCrZ9pjeqSU5Yn9LPu7umJQBIIv+k7vIOrtK66L4lmr3pVOrOv3otp619KZ0/KjdNI79L52Uo09FBQWrU0vfe5trezU+hLsoUKd3Repovte+0vbq/7Lj5XbaHECKasR9G4MPlbp+gzZxd6koPEJCkAYC5SjcOTAIIOK90Dja1IfIZ8Z+zAY9jm3b5Ia+MT5sFcqRJrR2AYYA8Kua5BzYRrFPNmD4PQMegGJMOffJJUsWiI3nCHZZjInNdffLWOufzbc3JaboCAVxwmnRHbhLSPwRJ4wU0BRSc6HkECYYVw95nMKgJOcylxrJttE5Ibzf9Xq9GPvP+WX3MiV/MGHfRu/SentRQrfG1GzsIrytdNXucSRKxQNIGHM9YhGFQJcdjNcBZvfJayuYe4Sia1CzwW+19mWOhe37HsxJWKwbu/jluEU15QzAQjAqCEbhMJc78GYV2E0kooHDubUImWkTOhGpgv8PoT8DJG/bzxna4BZ0eOFSOaLADGeSpFsg5AzeaDZIDQQXjZ4y/8ryfzUXBwdELRjTjCNvOeT0rNlrJz90vwy6N9pXXQEluX0inElpPWokSdiLCfiNJJjMKQ8Qsh8GEKQKMo/eiHrNbI9UksAAAAASUVORK5CYII="); }
```
Comment (jfclere) [inline wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: sure?

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+.colorpicker-hue.colorpicker-hidden,
+.colorpicker-bar.colorpicker-hidden {
+  display: none; }
+
+.colorpicker-inline.colorpicker-visible {
+  display: inline-block; }
+
+/** EXTENSIONS **/
+.colorpicker-preview {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAABkCAQAAAAVxWkcAAABr0lEQVR4Xo2VwU0DQQxF7dmRuNIFlzlSAR3QAaXQQdIBJVABFXDcOVAAd67cjJLR07dkhcSrkZKfb/t7bG88rFo3B5gZPMNycItu2xloGV7MWHzM9zuzFWCkmA0nK6AszCUJDW6+mG6R03ncw5v8EMTEvZ2O3AliYjpslblc0RF9LmZYWxURU6aKytWZYsoWCAe+xwOZp1GsEukGiIkYxcQCHck99+gRgB7JncyIB5SGEhP3Yh5P6JwX+u6AnYot104d8DJT7uH7M9JH6OZbimj0vfMVaYnJIZFJDBW9kHlerL2C6JV4mSt7uuo2N57RxnZ+usQjn0R1jwBJBrNO3evJpVYUWsJ/E3UiXRlv24/7YZ04xmEdWlzcKS+B/eapeyMvFd2k0+hRk/T0AmTW8h69s2sjYMsdPntECiILhAeIMZAeH4QvUwfn6ijC0tTV+fT9ky8jM9nK2g7Ly1VjSpKYq6IvsAm7MtNu1orEqa/K3KNvgMFdhfquPfJmp2dbh0/8Gzb6Y22ViaNr6n5410zXdngVhbu6XqdOtWOuin5hjABGp4a2uotZ71MVCfwDBt2/v37yo6AAAAAASUVORK5CYII=");
```
Comment (jfclere) [inline wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: sure?

File: wardemo/src/main/java/org/jfclere/tomcatPI/FrameBuffer.java (hunk 1 above)
Comment (gottburgm) [reply wardemo/src/main/java/org/jfclere/tomcatPI/FrameBuffer.java]: Well,  this probably is a better solution but they shouldn't be null at any time as the JavaScript which send the request is triggered only when a color (so a RGB value) is selected.

to be moresecure/clean i would change the following things in the **script.js** : 

  ```javascript

  function getRed(rgb_str) {
    var red = 0;
    var rgb_red = /rgba?\(\s*(\d+),(?:[\d\s]*),(?:[\d\s]*)(?:,\s*[\d\.]+)?\)/.exec(rgb_str);
    if(rgb_red) {
       red = rgb_red;
    }

    return red;
  }

  function getGreen(rgb_str) {
    var green = 0;
    var rgb_green = /rgba?\((?:[\d\s]*),\s*(\d+),(?:[\d\s]*)(?:,\s*[\d\.]+)?\)/.exec(rgb_str);
    if(rgb_green) {
       green = rgb_green;
    }

    return green;
  }

  function getBlue(rgb_str) {
    var blue = 0;
    var rgb_blue = /rgba?\((?:[\d\s]*),(?:[\d\s]*),\s*(\d+)(?:,\s*[\d\.]+)?\)/.exec(rgb_str);
    if(rgb_blue) {
       blue = rgb_blue;
    }

    return blue;
  }


  ```

  and in FrameBuffer.java :
  
  ```java
    if (param_j != null && param_i != null && param_red != null && param_green != null && param_blue != null) {
        int p_i = Integer.parseInt(param_i);
        int p_j = Integer.parseInt(param_j);
        int red = Integer.parseInt(param_red);
        int green = Integer.parseInt(param_green);
        int blue = Integer.parseInt(param_blue);

        if(p_i >= 0 && p_i <= 7 && p_j >= 0 && p_j <= 7 && pi.color(red, green, blue)) {
            pi.writepix(p_i, p_j, pi.color(red, green, blue));
        }
    }
  
  ```

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css (hunk 3 above)
Comment (gottburgm) [reply wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: Well as i didn't edited the file, and it comes directly from the official repository i don't think that we should change it.

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css (hunk 4 above)
Comment (gottburgm) [reply wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: Same as before for me too :)

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css (hunk 5 above)
Comment (gottburgm) [reply wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: again .

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css (hunk 7 above)
Comment (gottburgm) [reply wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: What about you ?

File: wardemo/src/main/webapp/css/bootstrap-colorpicker.css (hunk 8 above)
Comment (gottburgm) [reply wardemo/src/main/webapp/css/bootstrap-colorpicker.css]: OKAY: https://cdnjs.cloudflare.com/ajax/libs/bootstrap-colorpicker/2.5.2/css/bootstrap-colorpicker.css

File: wardemo/src/main/java/org/jfclere/tomcatPI/PIFrameBuffer.java (hunk 2 above)
Comment (gottburgm) [reply wardemo/src/main/java/org/jfclere/tomcatPI/PIFrameBuffer.java]: Was not sure about it, we can remove it !

File: wardemo/src/main/java/org/jfclere/tomcatPI/FrameBuffer.java (hunk 1 above)
Comment (jfclere) [reply wardemo/src/main/java/org/jfclere/tomcatPI/FrameBuffer.java]: +1

File: wardemo/src/main/java/org/jfclere/tomcatPI/PIFrameBuffer.java (hunk 2 above)
Comment (jfclere) [reply wardemo/src/main/java/org/jfclere/tomcatPI/PIFrameBuffer.java]: OK Cool remove it and update the PR.

File: wardemo/src/main/java/org/jfclere/tomcatPI/FrameBuffer.java [hunk 9]
```diff
@@ -115,9 +112,9 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
     }
 
 	/**
-	 * @see HttpServlet#doPost(HttpServletRequest request, HttpServletResponse response)
+	 * @see HttpServlet#doGet(HttpServletRequest request, HttpServletResponse response)
 	 */
-	protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+	protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
```
Comment (jfclere) [inline wardemo/src/main/java/org/jfclere/tomcatPI/FrameBuffer.java]: Sure you want an infinite loop there?

**Review by jfclere — APPROVED** [review]

**Pull request merged**